enhancements/namespace-auto-pruning.md

```diff
@@ -0,0 +1,170 @@
+* Allowing policy configurations at the repository level
+
+## Proposal
+
+The details in this proposal is only for namespace auto-pruning policies but designed in a way to support repository auto-pruning policies in the future. This proposal is structured in three parts: how policies are managed, how the policies are structured in the DB, and how the policies will be executed via the auto-prune worker.
+
+### Managing Policies
+
+A user is able to configure auto-pruning within the namespace settings. To start, there will be two auto-pruning options:
+* Prune by number of tags, provided the number of tags
+* Prune tags older than date, provided the number of days 
+
+For this implementation users will only be able to configure one policy per namespace. To interface with this new UI there will be a new set of CRUD endpoints for managing auto-pruning policies. The endpoints returning lists will be paginated.
+
+**Create policy**
+```
+POST /api/v1/organization/{organization}/autoprunepolicy
+By date:
+{
+    pruneBy: "creation_date",
+    olderThan: "2w"  
+}
+By number of tags:
+{
+    pruneBy: "number_of_tags",
+    numTags: 100  
+}
+```
+
+**Update policy**
+```
+PATCH /api/v1/organization/{organization}/autoprunepolicy/{policyid}
+{
+    fieldToBeUpdated: value,
+}
+```
+
+**List policies**
+```
+GET /api/v1/organization/{organization}/autoprunepolicy
+```
+
+**Get policy**
+```
+GET /api/v1/organization/{organization}/autoprunepolicy/{policyid}
+```
+
+**Delete policy**
+```
+DELETE /api/v1/organization/{organization}/autoprunepolicy/{policyid}
+```
+
+### Database structure
+Two tables will be required.
+
+**namespaceautoprunepolicy**
+
+This table holds the policy configuration for a single namespace. For this proposal there will be only one entry per namespace but there will be support for multiple rows per namespace_id. The config field will hold the policy details such as `{pruneBy: "creation_date", olderThan: "2w"}` or `{pruneBy: "number_of_tags",  numTags: 100}`.
+
+| Field | Datatype | Attributes | Description|
+| --- | ----------- | ----------- | ----------- |
+| uuid | character varying(255) | Unique, not null, indexed | Unique identifier for this policy |
+| namespace_id | integer | FK, not null | Namespace that the policy falls under |
+| config | text | JSON, not null | Policy configuration |
+
+**autoprunetask**
+
+This table registers tasks to be executed by the auto-prune worker. Tasks are executed within the context of a single namespace. Only one task per namespace will ever exist.
+
+| Field | Datatype | Attributes | Description |
+| --- | ----------- | ----------- | ----------- |
+| namespace_id | integer | FK, not null | Namespace that this task belongs too |
+| last_ran_ms | bigint | nullable, indexed | Last time the worker executed the policies for this namespace |
+| is_running | boolean | default false | Policies for this namespace are currently being executed |
```

Comment:
In the future when we start supporting multiple policies for a namespace, since neither of the cols are not unique, there can be a possibility of collisions. 
Can we have a FK, reference to `namespaceautoprunepolicy.uuid` in `autoprunetask`? So, we know which policy is running & can be useful when debugging too as to why a policy is stuck/did not run.

Reply:
The rows in this table are for running policies for a specific namespace and not a specific policy, so there would be no duplicate `namespace_id` values. When we support multiple namespace policies we'll fetch them all in [this step](https://github.com/quay/enhancements/pull/25/files#diff-355ab33b15f7b6f2ac5e26ca752cbd9d70fc3d9f60d03d84f70f55f47cf32c53R145) instead of just the one. We can't have a FK for a task because a task will eventually run multiple policies at once. For debugging info i'll add a status field as hank suggested that will have details if there's an error.

Reply:
Ah! ok, sounds good, I got it wrong.

Reply:
No problem!